farming/logics/traits/farming.rs

```diff
@@ -25,17 +30,112 @@ pub const MAX_PERIOD: u8 = 23u8;
 pub const FIRST_PERIOD_REWERD_SUPPLY: Balance = 151629858171523000000u128;
 
 #[openbrush::trait_definition]
-pub trait Farming: Storage<Data> + Storage<ownable::Data> + FarmingGetters {
+pub trait Farming: Storage<Data> + Storage<ownable::Data> + FarmingGetters + FarmingEvents {
     #[ink(message)]
     #[modifiers(only_owner)]
     fn add(
         &mut self,
-        alloc_point: u128,
+        alloc_point: u32,
         lp_token: AccountId,
-        rewarder: AccountId,
+        rewarder: Option<AccountId>,
     ) -> Result<(), FarmingError> {
         self._check_pool_duplicate(lp_token)?;
         self._update_all_pools()?;
+        self.data::<Data>().total_alloc_point = self
+            .data::<Data>()
+            .total_alloc_point
+            .checked_add(alloc_point)
+            .ok_or(FarmingError::AddOverflow2)?;
+        self.data::<Data>().lp_tokens.push(lp_token);
+        self.data::<Data>().rewarders.push(rewarder);
```

Comment:
save to storage only if Some()

Reply:
It was to align with proper pool id, but as you pointed out above, I'll change it using Hashmap structure.